### PR TITLE
feat(automl): redesign prediction type selection with recommendations

### PR DESCRIPTION
**Pull request by MatthewAThompson**
https://redhat.atlassian.net/browse/RHOAIUX-2589

## Description

Redesigns the AutoML configure **Prediction type** section to guide users based on the selected target column and file schema.

**What changed**
- Replaces the flat gallery of four prediction type cards with:
  - **Helper text** (`AutomlPredictionTypeHelperText`) shown as soon as a file is selected — indeterminate prompt before a target is chosen; after selection, a unique-value summary with a `?` popover.
  - **Recommended / not recommended cards** (`AutomlPredictionTypeSelector`) using PatternFly selectable compact cards (radio on the left), badges, descriptions, and italic recommendation reasons.
  - **Accordion** for “other prediction types” (plain-style, no filled background).
- Adds `predictionTypeUtils` for assessment logic:
  - Hard/soft incompatibility reasons (e.g. binary requires ≤2 values, multiclass requires ≥3, regression/timeseries rules for string targets).
  - Not-recommended types remain **selectable** (no disabled radios).
  - Inferred default (BFF `task_type` or timeseries when a timestamp column exists + numeric target) is ordered **first** among recommended tiles.
- Adds `unique_count` to the CSV schema API (BFF + OpenAPI + frontend `ColumnSchema`) for cardinality messaging.
- Extends timestamp detection with the `ds` column name (common forecasting datestamp).
- Removes `getColumnConstraintTooltip` in favour of the new assessment utilities.

**Why**
- Aligns the configure flow with updated UX/design: guided recommendations instead of equal-weight cards with disabled tooltips.
- Surfaces schema-derived context (unique values, timestamp column) to help users pick an appropriate prediction type.

## How Has This Been Tested?

https://github.com/user-attachments/assets/ffe6b77e-1048-432c-8a55-ece1bf1480f2

**Manual testing steps**
1. Open AutoML configure with a CSV file selected.
2. Confirm prediction type helper text appears before choosing a target column.
3. Select a **binary** target (2 unique values) — binary recommended; multiclass under “other prediction types”.
4. Select a **multiclass** target (3+ categories) — multiclass recommended; binary not recommended.
5. Select a **numeric** target with a `ds` (or other timestamp) column — timeseries recommended/auto-selected; timestamp column prefilled.
6. Select a **string** target — timeseries/regression not recommended but still selectable.
7. Expand/collapse “Show other prediction types” accordion; confirm plain background.

## Test Impact

- Updated `AutomlConfigure.spec.tsx` and `AutomlConfigurePage.spec.tsx` for the new selector, helper text, and recommendation behaviour.
- Added `predictionTypeUtils.spec.ts` (assessments, incompatibility rules, inferred default ordering, `ds` timeseries).
- Updated `columnUtils.spec.ts` (`ds` detection, `unique_count` helpers; removed deprecated tooltip tests).

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes:
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

## Summary by CodeRabbit

* **New Features**
  * Per-column unique value counts included in CSV schema and UI to improve cardinality-driven prediction recommendations
  * Redesigned prediction-type UI: selector, helper text, inferred/reordered recommendations, and clearer recommendation reasons
  * Improved timestamp detection (adds "ds") for better timeseries inference

* **Bug Fixes**
  * Whitespace-only target names now treated as unselected, restoring correct prediction-type visibility and reset behavior

* **Tests**
  * Expanded tests for prediction-type logic, helper text, inference, and unique-count handling

### CODE REVIEW

File: packages/automl/bff/internal/integrations/s3/client.go [hunk 1]
```diff
@@ -476,9 +477,10 @@ func (c *RealS3Client) GetCSVSchema(ctx context.Context, bucket, key string) (CS
 		uniqueValues := collectUniqueRawValues(dataRows, i)
 		taskType := inferTaskType(uniqueValues)
 		columnSchemas[i] = ColumnSchema{
-			Name:     colName,
-			Type:     inferColumnType(dataRows, i),
-			TaskType: taskType,
+			Name:        colName,
+			Type:        inferColumnType(dataRows, i),
+			TaskType:    taskType,
+			UniqueCount: len(uniqueValues),
```
[the remark this answers is not in the document]
Comment (MatthewAThompson) [reply packages/automl/bff/internal/integrations/s3/client.go:483]: Float64 dedup only collapses distinct numeric strings above ~2^53, which is outside typical AutoML column values, and exact-string dedup would break the intentional normalization of equivalent forms like "1" and "1.0".

File: packages/automl/frontend/src/app/components/configure/AutomlPredictionTypeSelector.scss [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+// Content small uses --pf-v6-c-content--small--Color (subtle grey); override with severity/warning tokens for not recommended reasons.
+.automl-prediction-type__reason--not-recommended {
+  --pf-v6-c-content--small--Color: var(--pf-t--global--color--severity--moderate--200);
+
+  :root:where(.pf-v6-theme-dark) & {
+    --pf-v6-c-content--small--Color: var(--pf-t--global--dark--color--severity--moderate--100);
+  }
+}
```
Comment (MatthewAThompson) [inline packages/automl/frontend/src/app/components/configure/AutomlPredictionTypeSelector.scss:31]: The finding is **not valid under this repo’s conventions**. Converting to CSS modules would go against established patterns, not fix an inconsistency.

**Package:** All 7 SCSS files in `packages/automl/frontend` are plain `.scss` imports (`import './Component.scss'`). There are no `.module.scss` files in the automl package.

**Repo:** `.claude/rules/css-patternfly.md` explicitly says:

> No CSS modules — use plain class imports: `import './MyComponent.scss';`

**This file already matches that pattern.** `AutomlPredictionTypeSelector.scss` uses the same approach as `AutomlConfigure.scss` — BEM-style prefixed classes (`.automl-prediction-type-*`) with nested `.pf-v6-*` overrides, which is the standard way to scope custom styles here.

**Remote leakage:** Module federation leakage is a real concern, but this package addresses it with prefixed class names, not CSS modules. That’s consistent with the rest of automl and the dashboard styling rules.

**Validation:** No code changes; existing SCSS/component wiring already follows package and repo conventions.

**Comment by nickmazzi** [comment]
Thanks @MatthewAThompson! This looks good. A few small comments from my pr review skill and the claude one.
The claude one is always a bit overkill. Feel free to only address the ones you thiink make sense. I already scrubbed many.

# AI assisted review
[pr-review-7890.md](https://github.com/user-attachments/files/28842288/pr-review-7890.md)

## Overall Assessment
**Approve with minor suggestions.** The code is well-structured, properly tested, and follows project conventions. A few items worth addressing below.


## Suggestions
2. **Accordion auto-expand UX**: Track whether the user explicitly collapsed the accordion to avoid re-expanding on unrelated re-renders.
3. **Remove `int64`/`float64` from `isNumericColumnType`**: If the TypeScript type union doesn't include them, they're dead branches. Either remove or document why they're kept.
4. **`unique_count` optionality**: The Go struct has `UniqueCount int` (zero-value 0), but the TS type has `unique_count?: number`. A column with truly 0 unique values would be indistinguishable from "not provided" on the TS side. Consider using a pointer in Go (`*int`) with `omitempty` if 0 is a meaningful value, or document that 0 means "empty column."

## pr-review-toolkit
These are some notes from the claude pr-review-toolkit. I snipped out irrelevant comments.

5. **Missing `isChecked` on selectable card actions** — [AutomlPredictionTypeSelector.tsx:76-86](packages/automl/frontend/src/app/components/configure/AutomlPredictionTypeSelector.tsx#L76-L86)
   The old code had `isChecked` in `selectableActions` for accessibility (screen readers announcing checked state). The new code omits it. Add `isChecked: isSelected` back.

7. **Dead code: `isNumericColumnType` checks `int64`/`float64` but Zod rejects them** — [predictionTypeUtils.ts:51](packages/automl/frontend/src/app/utilities/predictionTypeUtils.ts#L51)
   The Zod schema only accepts `integer | double | timestamp | bool | string`. The `int64`/`float64` branches in `isNumericColumnType` can never match validated data. Either add them to the Zod schema or remove the dead branches.

8. **`PredictionTypeAssessment` allows illegal states** — The `isRecommended` boolean + optional `notRecommendedReason` permits inconsistent combinations. A discriminated union (`isRecommended: true` without reason vs `isRecommended: false` with required reason) would enforce the invariant structurally.

9. **`PREDICTION_TYPE_OPTIONS` can drift from `PredictionTypeValue`** — Using a `Record<PredictionTypeValue, ...>` as the source would give a compile error when a new task type is added but no option is defined.

10. **`AutomlPredictionTypeHelperTextProps` has redundant props** — Both `targetColumn` (string) and `selectedColumn` (ColumnSchema) represent the same state. Collapsing to just `selectedColumn?: ColumnSchema` and using `selectedColumn?.name` would simplify.

14. **Minor gaps in utility function edge cases** — Missing tests for: `undefined` column input on all assessment functions, `uniqueCount == null` for binary, singular message for multiclass with 1 value, and `orderRecommendedAssessments` when preferred type is undefined/already first. These are low-risk but would improve coverage completeness.

16. **Accordion auto-expand UX** — Track whether the user explicitly collapsed the accordion to avoid re-expanding on unrelated re-renders.


# Test Validation

## automl

### Overall Summary

| Suite      | Result |
|------------|--------|
| Frontend   | ✅     |
| BFF        | ✅     |
| Contract   | ✅     |
| BFF Lint   | ✅     |
| Lint       | ✅     |
| Type Check | ✅     |

### Cypress
Validated. Also double checked and the data test ids which were removed were ported to the new form item

image


# Manual validation

LGTM!

image

File: packages/automl/frontend/src/app/utilities/predictionTypeUtils.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  [TASK_TYPE_TIMESERIES]: {
+    label: TASK_TYPE_LABELS[TASK_TYPE_TIMESERIES],
+    description: 'Predict future values based on time-ordered historical data.',
+  },
+} satisfies Record<PredictionTypeValue, { label: string; description: string }>;
+
+export const PREDICTION_TYPE_OPTIONS: PredictionTypeOption[] = TASK_TYPES.map((taskType) => ({
+  value: taskType,
+  ...PREDICTION_TYPE_CONFIG[taskType],
+}));
```
[the remark this answers is not in the document]
Comment (MatthewAThompson) [reply packages/automl/frontend/src/app/utilities/predictionTypeUtils.ts:42]: `PREDICTION_TYPE_CONFIG` already uses `satisfies Record<PredictionTypeValue, …>`, and `PredictionTypeValue` is derived from `TASK_TYPES` via the configure schema, so a missing or extra option fails at compile time. Building `PREDICTION_TYPE_OPTIONS` from `TASK_TYPES` keeps option order aligned with the form’s canonical task-type list; switching to `Object.entries(PREDICTION_TYPE_CONFIG)` wouldn’t add safety and could let display order drift from `TASK_TYPES`.

**Comment by nickmazzi** [comment]
Thanks for updates @MatthewAThompson!

I tested again and looks good. I ran cypress locally and multiclass/regression/time series passed. I hit an error on upload for binary but I think that’s unrelated to your changes.

/lgtm
/approve

EDIT
I reran binary and it worked as well 🙏
